tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,263 @@
+    Program	Type	3	18	20	Value	=	TAN	(TANF)	or	Value	=	SSP	(SSP-MOE)
+    Edit Indicator	1	21	21	1=Return	Fatal	&	Warning	Edits	2=Return	Fatal	Edits	only
+    Encryption      1	22	22	E=SSN	is	encrypted	Blank	=	SSN	is	not	encrypted
+    Update      	1	23	23	N	=	New	data	D	=	Delete	existing	data	U
+    QUARTERS:
+        Q=1	(Jan-Mar)
+        Q=2	(Apr-Jun)
+        Q=3	(Jul-Sep)
+        Q=4	(Oct-Dec)
+    Example:
+    HEADERYYYYQTFIPSSP1EN
+    """
+
+    logger.debug('Validating header row.')
+    section_map = {
+        'A': 'Active Case Data',
+        'C': 'Closed Case Data',
+        'G': 'Aggregate Data',
+        'S': 'Stratum Data',
+    }
+
+    try:
+        header = {
+            'title':        row[0:6],
+            'year':         row[6:10],
+            'quarter':      row[10:11],
+            'type':         row[11:12],
+            'state_fips':   row[12:14],
+            'tribe_code':   row[14:17],
+            'program_type': row[17:20],
+            'edit':         row[20:21],
+            'encryption':   row[21:22],
+            'update':       row[22:23],
+        }
+
+        for key, value in header.items():
+            logger.debug('Header key %s: "%s"' % (key, value))
+
+        # TODO: Will need to be saved in parserLog, #1354
+
+        try:
+            logger.debug("Given section: '%s'\t Header section: '%s'", given_section, section_map[header['type']])
+            logger.debug("Given program type: '%s'\t Header program type: '%s'", data_type, header['program_type'])
+
+            if given_section != section_map[header['type']]:
+                raise ValueError('Given section does not match header section.')
+            if (data_type == 'TANF' and header['program_type'] != 'TAN')\
+                    or (data_type == 'SSP' and header['program_type'] != 'SSP')\
+                    or (data_type not in ['TANF', 'SSP']):
+                raise ValueError("Given data type does not match header program type.")
+        except KeyError as e:
+            logger.error('Ran into issue with header type: %s', e)
+
+        # TODO: could import schema from a schemas folder/file, would be reusable for other sections
+        # ugh, seems like we're limited to string for all values otherwise we have to cast to int in header and it fails
+        # with a raised ValueError instead of in the validator.errors
+        header_schema = {
+            'title':        {'type': 'string', 'required': True, 'allowed': ['HEADER']},
+            'year':         {'type': 'string', 'required': True, 'regex': '^20[0-9]{2}$'},
+            'quarter':      {'type': 'string', 'required': True, 'allowed': ['1', '2', '3', '4']},
+            'type':         {'type': 'string', 'required': True, 'allowed': ['A', 'C', 'G', 'S']},
+            'state_fips':   {'type': 'string', 'required': True, 'regex': '^[0-9]{2}$'},
+            'tribe_code':   {'type': 'string', 'required': False, 'regex': '^([0-9]{3}|[ ]{3})$'},
```

Comment:
All STTs will have a state_fips code -- tribes have their parent state. For States, tribe_code will be an empty "   ". I believe Territories just have their own `state_fips`.